ros/ros2/OdometryServer.cpp

```diff
@@ -80,111 +80,94 @@ OdometryServer::OdometryServer(const rclcpp::NodeOptions &options)
     if (publish_odom_tf_) {
         geometry_msgs::msg::TransformStamped transform_msg;
-        transform_msg.header.stamp = msg->header.stamp;
-        transform_msg.header.frame_id = odom_frame_;
-        transform_msg.child_frame_id = child_frame_;
-        transform_msg.transform.rotation.x = q_current.x();
-        transform_msg.transform.rotation.y = q_current.y();
-        transform_msg.transform.rotation.z = q_current.z();
-        transform_msg.transform.rotation.w = q_current.w();
-        transform_msg.transform.translation.x = t_current.x();
-        transform_msg.transform.translation.y = t_current.y();
-        transform_msg.transform.translation.z = t_current.z();
+        transform_msg.header = odom_header;
+        transform_msg.child_frame_id = egocentric_estimation ? cloud_frame_id : base_frame_;
+        transform_msg.transform = tf2::sophusToTransform(pose);
         tf_broadcaster_->sendTransform(transform_msg);
     }
 
     // publish trajectory msg
     geometry_msgs::msg::PoseStamped pose_msg;
-    pose_msg.pose.orientation.x = q_current.x();
-    pose_msg.pose.orientation.y = q_current.y();
-    pose_msg.pose.orientation.z = q_current.z();
-    pose_msg.pose.orientation.w = q_current.w();
-    pose_msg.pose.position.x = t_current.x();
-    pose_msg.pose.position.y = t_current.y();
-    pose_msg.pose.position.z = t_current.z();
-    pose_msg.header.stamp = msg->header.stamp;
-    pose_msg.header.frame_id = odom_frame_;
+    pose_msg.header = odom_header;
+    pose_msg.pose = tf2::sophusToPose(pose);
     path_msg_.poses.push_back(pose_msg);
     traj_publisher_->publish(path_msg_);
 
     // publish odometry msg
     auto odom_msg = std::make_unique<nav_msgs::msg::Odometry>();
-    odom_msg->header = pose_msg.header;
-    odom_msg->child_frame_id = child_frame_;
+    odom_msg->header = odom_header;
     odom_msg->pose.pose = pose_msg.pose;
     odom_publisher_->publish(std::move(odom_msg));
 
-    // Publish KISS-ICP internal data, just for debugging
-    auto frame_header = msg->header;
-    frame_header.frame_id = child_frame_;
-    frame_publisher_->publish(std::move(EigenToPointCloud2(frame, frame_header)));
-    kpoints_publisher_->publish(std::move(EigenToPointCloud2(keypoints, frame_header)));
-
-    // Map is referenced to the odometry_frame
-    auto local_map_header = msg->header;
-    local_map_header.frame_id = odom_frame_;
-    map_publisher_->publish(std::move(EigenToPointCloud2(odometry_.LocalMap(), local_map_header)));
+    // Publish KISS-ICP internal map, just for debugging
+    if (map_publisher_->get_subscription_count() > 0) {
+        const auto kiss_map = odometry_.LocalMap();
+        if (egocentric_estimation) {
+            map_publisher_->publish(std::move(EigenToPointCloud2(kiss_map, odom_header)));
+        } else {
+            const auto T = CloudToBaseTf(cloud_frame_id).inverse();
+            map_publisher_->publish(std::move(EigenToPointCloud2(kiss_map, T, odom_header)));
```

Comment:
Issue https://github.com/PRBonn/kiss-icp/issues/243 also applies here. Since the `kiss_map` is expressed w.r.t. the first LiDAR frame and we want to express it w.r.t. the first base frame, we have to use the "first" `CloudToBaseTf` when the first LiDAR frame was processed. This does of course not matter if the tf between sensor and base is static.

Reply:
Updated the issue so this doesn't get lost